### ISSUE TITLE
[Permissions] Replace `token` with `credentials` on the `PermissionsServiceRequestOptions` type

### ISSUE THREAD

**Issue by camilaibs** [backstage/backstage#25136]
### Context
Ref: https://github.com/backstage/backstage/issues/24493
The service definitions are being reviewed as we prepare for the stable release of the New Backend System.

### Description
Credentials for the permissions service request options should be used in place of the token, the token needs to go by the time of the stable release.

### Acceptance criteria
Deprecate the previous type and re-export it from a `deprecated.ts` file.

**Comment by camilaibs** [comment]
The `permissions` service types appear to be nicely prefixed, following the pattern that was used for renaming URLReader service types [here](https://github.com/backstage/backstage/pull/24883/files#diff-d7bdc8edbf9afc06219cd1284137b049dab795e0e812947b608dc8ab517c6623).

The `token` is already deprecated, so do you think we can put this issue on hold so the token can be removed later along with the release pull request @Rugvip ?

**Comment by Rugvip** [comment]
Hmm, we want to remove the imports from permission-common too, for example AuthorizePermissionRequest

**Comment by camilaibs** [comment]
> Hmm, we want to remove the imports from permission-common too, for example AuthorizePermissionRequest

Could you please remind me why we want to remove imports from permission-common?

**Comment by flash-me** [comment]
Hey @camilaibs, 
may you confirm, if I understand this issue in a correct way? 

As of now, the `handle`-Method of a [`PermissionPolicy`](https://github.com/backstage/backstage/blob/e43d17bea734b6cb271419093c6a0865376c186c/plugins/permission-node/src/policy/types.ts#L53-L58) gets an optional `BackstageIdentityResponse` passed, which in turn contains a token that has been [created and passed here](https://github.com/backstage/backstage/blob/master/plugins/permission-backend/src/service/router.ts#L133-L153).

With this issue, not a token is passed anymore, but the credential, correct?
![image](https://github.com/backstage/backstage/assets/5240450/09e8668e-fc0f-4bd6-bfd3-4de8139bbca7)

Best regards
A ⚡️

**Comment by Rugvip** [comment]
@flash-me Two main options I think is to either add `credentials` to the existing object, i.e. `{ credentials, identity, token }`, or to make the credentials the new `user` object, i.e. `{ ...credentials, identity, token }`. Either way we'll need to keep both `identity` and `token` around to make sure that we don't break existing policies, although we'll want to deprecate both of them since the `credentials` object contains the same information. Tbh I think the first option is the best, it's less risky as there might be conflicts in the `credentials` object (it has a `token` internally in some cases). It's also is a bit better for the future in case we want to provide additional metadata for the calling identity in addition to the `credentials` specifically for permission handlers.

With the above suggestion it's gonna make sense to create a new type in the permission plugin that mimics the `BackstageIdentityResponse`, but adds `credentials` and deprecates the old fields. Maybe call it something like `PermissionPolicyHandlerIdentity`?

**Comment by flash-me** [comment]
@Rugvip Sounds sensible to me, as it is always complicated to change API's once they went public 🫣

But I cannot agree with the following, which is also the reason for my curiosity:

>  since the credentials object contains the same information

This is not true (Maybe only for us??!!). The token in the `credentials` object is the Bearer Token what we get from our Identity Provider, which contains additional claims (and what we need!).

The other `token`, which is currently passed to the permissions through the `handle` method is a Backstage Token that **misses all the claims we receive from our IDP.** (the result of `getPluginRequestToken`)

Right now we have to workaround this by implementing an express Middleware on the /authorize Endpoint that stores the Bearer Token from our IDP in a cache, which we then get it again within the `handle` method of the PermissionPolicy, before it is replaced with the `BackstageToken`. And this feels like it is not the intended way.  😵‍💫

Best regards
Adem

**Comment by Rugvip** [comment]
@flash-me ah, let me rephrase that a bit: the credentials object *should* contain the same information. Either directly or indirectly by passing the credentials to other services, the same information should be accessible. That's not the case for custom claims though yet, but it's something we're looking to solve broadly for credentials, so we can count on that separate change happening at some point.

**Comment by flash-me** [comment]
@Rugvip thank you for coming back and for the insights! Appreciate it!

Best
A

**Comment by camilaibs** [comment]
> @Rugvip thank you for coming back and for the insights! Appreciate it!
> 
> Best A

@flash-me, would you be interested in contributing? We would be happy to provide feedback in a pull request if you are interested in contributing. Let me know if you would like me to assign you the issue 🙂 .

**Comment by camilaibs** [comment]
@flash-me it is only related to the permissions service, it's a subtask of https://github.com/backstage/backstage/issues/24493, which contains a series of tasks related to refining the new backend system framework and core plugins before the stable release of the new backend system.

**Comment by Rugvip** [comment]
Proposed updates to the policy in #25604 as well
